File: orchestrator/app/agents/budget.py
import re


_BUDGET_PATTERN = re.compile(
    r"(?:monthly\s+budget|月预算|每月预算)\D*?(\d+(?:\.\d+)?)",
    re.IGNORECASE,
)


def extract_monthly_budget(context: str) -> float | None:
    matches = _BUDGET_PATTERN.findall(context or "")
    if not matches:
        return None
    try:
        budget = float(matches[-1])
    except ValueError:
        return None
    return budget if budget > 0 else None


def format_monthly_budget(budget: float) -> str:
    return f"MONTHLY BUDGET: ${budget:.2f}"

File: web/src/lib/budget.ts
const createMonthlyBudgetPattern = () =>
  /(?:monthly\s+budget|月预算|每月预算)\D*?(\d+(?:\.\d+)?)/i;

export const parseMonthlyBudget = (context: string): number | null => {
  const matches = context.match(
    new RegExp(createMonthlyBudgetPattern().source, "gi")
  );
  if (!matches) return null;
  const latest = matches[matches.length - 1];
  const match = latest.match(createMonthlyBudgetPattern());
  if (!match) return null;
  const budget = Number(match[1]);
  return Number.isFinite(budget) && budget > 0 ? budget : null;
};

export const formatMonthlyBudget = (budget: number): string =>
  `MONTHLY BUDGET: $${budget.toFixed(2)}`;

export const replaceMonthlyBudget = (
  context: string,
  budget: number
): string => {
  const budgetLine = formatMonthlyBudget(budget);
  if (!context.trim()) return budgetLine;
  return createMonthlyBudgetPattern().test(context)
    ? context.replace(
        new RegExp(createMonthlyBudgetPattern().source, "gi"),
        budgetLine
      )
    : `${context.trimEnd()} ${budgetLine}`;
};
